src/mgr/DaemonServer.cc

```diff
@@ -1766,6 +1766,182 @@ void DaemonServer::send_report()
+		  dout(10) << "pool " << i.first
+			   << " target " << p.get_pg_num_target()
+			   << " pg_num " << p.get_pg_num()
+			   << " - no state for " << merge_source
+			   << " (merge source)"
+			   << dendl;
+		  ok = false;
+		} else if (!(q->second.state & (PG_STATE_ACTIVE |
+						PG_STATE_CLEAN))) {
+		  dout(10) << "pool " << i.first
+			   << " target " << p.get_pg_num_target()
+			   << " pg_num " << p.get_pg_num()
+			   << " - merge source " << merge_source
+			   << " not clean (" << pg_state_string(q->second.state)
+			   << ")" << dendl;
+		  ok = false;
+		}
+		q = pg_map.pg_stat.find(merge_target);
+		if (q == pg_map.pg_stat.end()) {
+		  dout(10) << "pool " << i.first
+			   << " target " << p.get_pg_num_target()
+			   << " pg_num " << p.get_pg_num()
+			   << " - no state for " << merge_target
+			   << " (merge target)"
+			   << dendl;
+		  ok = false;
+		} else if (!(q->second.state & (PG_STATE_ACTIVE |
+						PG_STATE_CLEAN))) {
+		  dout(10) << "pool " << i.first
+			   << " target " << p.get_pg_num_target()
+			   << " pg_num " << p.get_pg_num()
+			   << " - merge target " << merge_target
+			   << " not clean (" << pg_state_string(q->second.state)
+			   << ")" << dendl;
+		  ok = false;
+		}
+		if (ok) {
+		  unsigned target = p.get_pg_num() - 1;
+		  dout(10) << "pool " << i.first
+			   << " target " << p.get_pg_num_target()
+			   << " pg_num " << p.get_pg_num()
+			   << " -> " << target
+			   << " (merging " << merge_source
+			   << " and " << merge_target
+			   << ")" << dendl;
+		  pg_num_to_set[osdmap.get_pool_name(i.first)] = target;
+		}
+	      } else if (p.get_pg_num_target() > p.get_pg_num()) {
+		// pg_num increase (split)
+		bool active = true;
+		auto q = pg_map.num_pg_by_pool_state.find(i.first);
+		if (q != pg_map.num_pg_by_pool_state.end()) {
+		  for (auto& j : q->second) {
+		    if ((j.first & (PG_STATE_ACTIVE|PG_STATE_PEERED)) == 0) {
+		      dout(20) << "pool " << i.first << " has " << j.second
+			       << " pgs in " << pg_state_string(j.first)
+			       << dendl;
+		      active = false;
+		    }
+		  }
+		} else {
+		  active = false;
+		}
+		if (!active) {
+		  dout(10) << "pool " << i.first
+			   << " target " << p.get_pg_num_target()
+			   << " pg_num " << p.get_pg_num()
+			   << " - not all pgs active"
+			   << dendl;
+		} else {
+		  unsigned add = std::min(
+		    left,
+		    p.get_pg_num_target() - p.get_pg_num());
+		  unsigned target = p.get_pg_num() + add;
+		  left -= add;
+		  dout(10) << "pool " << i.first
+			   << " target " << p.get_pg_num_target()
+			   << " pg_num " << p.get_pg_num()
+			   << " -> " << target << dendl;
+		  pg_num_to_set[osdmap.get_pool_name(i.first)] = target;
+		}
+	      }
+	    }
+
+	    // adjust pgp_num?
+	    unsigned target = std::min(p.get_pg_num_pending(),
+				       p.get_pgp_num_target());
+	    if (target != p.get_pgp_num()) {
+	      // FIXME: we should throttle this to limit mispalced objects, like
```

Comment:
FIXME - is this fixed later?

Reply:
since we're sending these each tick, but only getting updated pg maps at a longer interval, won't this duplicate effort?

Reply:
nope, this is a future TODO.  right now teh user can jump to any pgp_num value so there is no throttling.  eventually, we should throttling pg increases over time the way the balancer does.

this might duplicate some effort, yeah.  it keeps the code really simple and stateless, though.  i'm not quite sure what throttling would make sense here.